openquake/hazardlib/gsim/sgobba_2020.py

```diff
@@ -0,0 +1,296 @@
+    #: this GSIM can calculate. A set should contain classes from module
+    #: :mod:`openquake.hazardlib.imt`.
+    DEFINED_FOR_INTENSITY_MEASURE_TYPES = {
+        PGA,
+        SA
+    }
+
+    #: Supported intensity measure component is the geometric mean of two
+    #: horizontal components
+    DEFINED_FOR_INTENSITY_MEASURE_COMPONENT = const.IMC.AVERAGE_HORIZONTAL
+
+    #: Supported standard deviation types are inter-event, intra-event
+    #: and total
+    DEFINED_FOR_STANDARD_DEVIATION_TYPES = {
+        const.StdDev.TOTAL,
+        const.StdDev.INTER_EVENT,
+        const.StdDev.INTRA_EVENT
+    }
+
+    #: Required site parameter is not set
+    REQUIRES_SITES_PARAMETERS = set()
+
+    #: Required rupture parameters is magnitude, ev_lat, ev_lon
+    REQUIRES_RUPTURE_PARAMETERS = {'mag', 'ev_lat', 'ev_lon'}
+
+    #: Required distance measure is Rjb
+    REQUIRES_DISTANCES = {'rjb'}
+
+    def __init__(self, event_id=None, directionality=False, cluster=None,
+                 site=False, bedrock=False, **kwargs):
+        super().__init__(event_id=event_id,
+                         directionality=directionality,
+                         cluster=cluster,
+                         **kwargs)
+        self.event_id = event_id
+        self.directionality = directionality
+        self.cluster = cluster
+        self.site = site
+        self.bedrock = bedrock
+        # Get site indexes. They are used for the site correction and the
+        # cluster (path) correction
+        # Load the coordinate of the grid
+        fname = os.path.join(DATA_FOLDER, 'grid.csv')
+        coo = np.fliplr(np.loadtxt(fname, delimiter=";"))
+        self.kdt = cKDTree(coo)
+
+        # Load the table with the between event coefficients
+        if event_id is not None:
+            self.event_id = event_id
+            fname = os.path.join(DATA_FOLDER, 'event.csv')
+            # Create dataframe with events
+            df = pd.read_csv(fname, dtype={'id': str})
+            df.set_index('id', inplace=True)
+            self.df = df
+            assert event_id in df.index.values
+
+    def get_mean_and_stddevs(self, sites, rup, dists, imt, stddev_types):
+        """
+        Eq.1 - page 2
+        """
+        # Ergodic coeffs
+        C = self.COEFFS[imt]
+        # between-event
+        if self.event_id is not None:
+            label = "dBe_{:s}".format(imt.__str__())
+            self.be = self.df.loc[self.event_id][label]
+            self.be_std = 0.0
+        else:
+            self.be_std = C['tau_ev']
+            self.be = 0.0
+        # Site correction
+        tmp = [[s.location.longitude, s.location.latitude] for s in sites]
+        dsts, self.idxs = self.kdt.query(np.array(tmp))
+        sc = 0
+        phi_S2Sref = C['phi_S2S_ref']
+        Bs_model = np.zeros(sites.vs30.shape)
+        if self.site and self.bedrock is False:
+            sc, Bs_model, phi_S2Sref = self._get_site_correction(sites.vs30.shape, imt)
```

Comment:
Actually I see that this implementation will always be slow because there are several geospatial operations for each ruptures. Perhaps the solution is to add in the docstring a big warning such as

**This GMPE is not meant for national models where it would be too slow to be practical, it should be used only in scenario calculations**

Reply:
Done, there are still some files uploaded in the "get_cluster_correction", but I think they cannot be moved in the init.